Adds RepositoryEditor

zmrow opened the pull request:
*This PR builds on the [`KeySource` trait PR](https://github.com/awslabs/tough/pull/103) that needs a Rusoto release to be merged. Once that PR is merged we can merge this one*

*Issue #, if available:*
Related to #43 
Builds on #103 

*Description of changes:*
This commit adds a RepositoryEditor struct, which allows a user
to build a new repository from scratch or edit an existing
`tough::Repository` using a pseudo builder-style pattern.

Users supply a path to a valid `root.json`, add targets, versions, and expiration dates, and then call `sign()` with signing keys. Provided all goes well, a user is given a `SignedRepository` struct which they can then use to write to disk. Convenience methods are provided to symlink targets to a location provided they exist in the `SignedRepository` and the checksums match.

`SignedRepository` does not keep track of locations of given targets. This responsibility is left to the calling code.

*Considerations and questions:*
* `RepositoryEditor` currently provides no facilities to manipulate versions or expirations; they must be provided. I think it would be nice to do so, but it's difficult to know what users could be using for versions and how they would want to increment them. Expiration dates could (should?) be unique for each role, so I didn't want to blanket implement something when a user should either update them, or save the current date and use it in the builder methods.

* While this abstracts away most of the logic from `tuftool` it does not help a user sign a *single* role (timstamp, targets, etc). We could make this possible by moving some of the logic into a `SignedRole.from_role(<ROLE HERE>)` method? I'm not sure how much of this we want to expose. It's also a bit weird to have `SignedRole` own this logic.

*Testing:*
* All unit tests pass and clippy is happy
* Added a few simple unit tests
* Currently working on an "integration" style test that (hopefully) can "sign" a repo, and then test the symlink methods on that signed repo.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

zmrow committed:
Adds RepositoryEditor

This commit adds a RepositoryEditor struct, which allows a user
to build a new repository from scratch or edit an existing
`tough::Repository` using a pseudo builder-style pattern

zmrow commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+    /// if the filename exists in `Targets`, the file's sha256 is compared
+    /// against the data in `Targets`. If this data does not match, the
+    /// method will fail. If all is well, the taret is symlinked.
+    pub fn link_targets<P1, P2>(&self, indir: P1, outdir: P2) -> Result<()>
```
I realize this method is fairly long and I spent entirely too long trying to break it up.  Given the data we need to keep track of and the order of operations here, I left the logic as is since it is straightforward and linear.

tjkirch commented on Cargo.lock:
```diff
@@ -1865,6 +1865,7 @@ dependencies = [
  "tempfile 3.1.0 (registry+https://github.com/rust-lang/crates.io-index)",
  "untrusted 0.7.1 (registry+https://github.com/rust-lang/crates.io-index)",
  "url 2.1.1 (registry+https://github.com/rust-lang/crates.io-index)",
+ "walkdir 2.3.1 (registry+https://github.com/rust-lang/crates.io-index)",
```
This seems odd; you have walkdir 2.2.9 in Cargo.toml, but we're getting 2.3.x here?  I would have expected it to pull in 2.2.latest by Cargo's default caret requirement rules.  Any idea why?

I think it'd make sense to use 2.2.x to match existing crates, or update everything to 2.3.x (probably separately) so they don't have to build twice, but I just wanted to raise this question...

zmrow replied:
No idea what's happening here. I even went so far as to recreate the `Cargo.lock` file but it still wants to pull in 2.3.1?

tjkirch replied:
What if you use an equals requirement instead of a (default) caret requirement?  (We don't actually want to do that, but it'd be nice to understand.  This isn't a blocker.)

tjkirch replied:
I misunderstood caret requirements - `x.y.z` allows `x.*.*` higher versions, not just `x.y.*` higher versions, if the dependency is past 1.0.  I guess I'm used to 0.x dependencies :)

https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#caret-requirements

tjkirch commented on tough/src/schema/mod.rs:
```diff
@@ -120,6 +123,18 @@ impl Root {
+
+    /// Given an object/key that impls Sign, return the corresponding
+    /// key ID from Root
+    #[allow(clippy::unnecessary_operation)]
+    pub fn key_id(&self, key_pair: &dyn Sign) -> Option<Decoded<Hex>> {
+        for (key_id, key) in &self.keys {
+            if key_pair.tuf_key() == *key {
+                Some(key_id.clone());
+            }
+        }
+        None
```
Is this clippy warning because we're not actually `return`ing the cloned `key_id`?  It seems like it'll always return `None`.

zmrow replied:
:)

Yes - that little `return` fixed all my testing issues as well. :)

webern replied:
Clippy can be pretty helpful at times.
Another one I've found is if I'm calling `fn do_something() -> Result<()>` like this `do_something();` clippy helpfully tells me I'm ignoring the error!

tjkirch commented on tough/tests/repo_editor.rs:
```diff
@@ -0,0 +1,60 @@
+use tough::editor::RepositoryEditor;
+use tough::{ExpirationEnforcement, FilesystemTransport, Limits, Repository, Settings};
+
+mod test_utils;
```
It might be clearer to have a `tough/tests/mod.rs` with `mod test_utils;` so each tests file doesn't have to declare it?

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Why split the tests between this file and the separate tests file?

zmrow replied:
I guess I'm drawing a line, albeit a squiggly one, between the unit tests and tests that include other parts of the library.

I was thinking that the tests that create/write/load would go here since they use more than just the `Editor` library. Tests that specifically test the `Editor` library would go in the lib.

tjkirch commented on tough/tests/repo_editor.rs:
```diff
@@ -0,0 +1,60 @@
+    let root = repo_paths.root_path.clone();
+    let repo = load_tuf_reference_impl(&mut repo_paths);
+
+    assert!(RepositoryEditor::from_repo(&root, repo).is_ok());
```
It seems like it'd be good to extend the test (or add a test) to change a field, `write` the repo, and then make sure the written repo also loads correctly through the standard `Repository::load`.

webern replied:
> It seems like it'd be good to extend the test (or add a test) to change a field, `write` the repo, and then make sure the written repo also loads correctly through the standard `Repository::load`.

I agree. If not this PR then maybe in the future, but more tests that serve the dual purpose of use case examples would be nice.

zmrow replied:
Yep - I have a separate integration test waiting in the wings for just that.

tjkirch commented on tough/src/error.rs:
```diff
@@ -18,6 +19,14 @@ pub type Result<T> = std::result::Result<T, Error>;
 #[derive(Debug, Snafu)]
 #[snafu(visibility = "pub(crate)")]
 pub enum Error {
+    /// The library failed to canonicalize the given path
```
```suggestion
```
I think the `display` is pretty clear - does this help in some other way?  I thought docstrings were mainly used if you don't need any variables in the `display`.  (Same with below docstrings)

zmrow replied:
Honestly I only added these to stay consistent with the style in the file.

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+        P1: AsRef<Path>,
+        P2: AsRef<Path>,
+    {
+        std::fs::create_dir_all(&outdir).context(error::DirCreate {
```
```suggestion
        let outdir = outdir.as_ref();
        std::fs::create_dir_all(&outdir).context(error::DirCreate {
```
`outdir` is used a number of times here; this will let you remove the `&` and `.as_ref()` when using it.

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+            path: indir.as_ref().to_owned(),
+        })?;
+        let abs_outdir = std::fs::canonicalize(&outdir).context(error::AbsolutePath {
+            path: outdir.as_ref().to_owned(),
```
```suggestion
            path: indir.as_ref(),
        })?;
        let abs_outdir = std::fs::canonicalize(&outdir).context(error::AbsolutePath {
            path: outdir.as_ref(),
```
Snafu should handle cloning for you, and only in the error case; this would clone even in the success case.  (If you need to do it manually for some reason I'm missing, `.with_context(|| bla)` will only run in the error case, which is good if the error creation takes a few cycles...)

tjkirch commented on tough/src/error.rs:
```diff
@@ -128,6 +167,21 @@ pub enum Error {
     },
 
+    /// A required reference to a metadata file is missing from a metadata file.
+    #[snafu(display("Missing field: {}", field))]
```
```suggestion
    #[snafu(display("Missing '{}' when building repo from RepositoryEditor", field))]
```

tjkirch commented on tough/src/error.rs:
```diff
@@ -128,6 +167,21 @@ pub enum Error {
+    /// Can't create a repo with no targets
+    #[snafu(display("Unable to create a TUF repo without targets"))]
+    NoTargets {},
+
```
```suggestion
```
Wound up not using this.

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+        if let Some(ref new_targets) = self.new_targets {
+            targets.extend(new_targets.clone());
+        }
+        //ensure!(!targets.is_empty(), error::NoTargets);
```
```suggestion
```

tjkirch commented on tough/src/error.rs:
```diff
@@ -128,6 +167,21 @@ pub enum Error {
+    #[snafu(display("Unable to create a TUF repo without targets"))]
+    NoTargets {},
+
+    #[snafu(display("Key for role '{}' do not exist in root.json", role))]
```
```suggestion
    #[snafu(display("Key for role '{}' doesn't exist in root.json", role))]
```

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+        let mut ser = serde_json::Serializer::with_formatter(&mut data, CanonicalFormatter::new());
+        role.signed
+            .serialize(&mut ser)
+            .context(error::SerializeRole)?;
```
Can we include the `T::TYPE` name in this error?  Same for SerializeSignedRole just below?

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+        // means that `entry.path()` call will return its absolute path.
+        let walker = WalkDir::new(abs_indir).follow_links(false);
+        for entry in walker {
+            let entry = entry.context(error::WalkDir)?;
```
Can we include the path we failed to walk in the error?

tjkirch reviewed (commented):
Going to review in a couple batches; this is everything but editor.rs.

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+
+        let filename = match T::TYPE {
+            RoleType::Targets => {
+                if consistent_snapshot {
+                    format!("{}.targets.json", self.signed.signed.version())
```
We have this naming logic in 4 places now, maybe we should do it in one place?  (tuftool/tests/refresh_command, Repository::*_filename, lib::load_targets, and now here)

webern replied:
I agree. It's probably also time to put those strings somewhere as consts.

zmrow replied:
https://github.com/awslabs/tough/issues/126

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+                if consistent_snapshot {
+                    format!("{}.snapshot.json", self.signed.signed.version())
+                } else {
+                    "targets.json".to_string()
```
```suggestion
                    "snapshot.json".to_string()
```

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+/// A set of signed TUF Repository metadata.
+///
+/// This metadata represents a signed TUF repository, but without the actual
+/// targets, the repository cannot be used. It is up to ther user to ensure
```
```suggestion
/// targets, the repository cannot be used. It is up to the user to ensure
```

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+/// targets, the repository cannot be used. It is up to ther user to ensure
+/// all the target files are in the proper directories along with the
+/// contained metadata. There are convenience methods provided to help with
+/// this.
```
```suggestion
/// This metadata represents a signed TUF repository and provides the ability to write the metadata to disk.
///
/// Note: without the target files, the repository cannot be used. It is up to the user to ensure
/// all target files referenced by the metadata are available. There are convenience methods to help with this.
```
Slight simplification, and not presenting a "negative" first.

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+
+        // Walk the absolute path of the indir. Using the absolute path here
+        // means that `entry.path()` call will return its absolute path.
+        let walker = WalkDir::new(abs_indir).follow_links(false);
```
Why not follow links?  I'm not sure we should dictate how someone lays out their directories.

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+            let entry = entry.context(error::WalkDir)?;
+
+            // If the entry is a file...
+            if entry.file_type().is_file() {
```
```suggestion
            if !entry.file_type().is_file() {
                continue;
            }
```
Rather than the `else` way below.

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+
+                // use the file name to see if a target exists in the repo
+                // with that name. If so...
+                if let Some(repo_target) = repo_targets.get(file_name) {
```
I wouldn't always recommend this, but there's so much nesting in the function that I think this would be clearer overall:
```suggestion
                let repo_target = match repo_targets.get(file_name) {
                    Some(repo_target) => repo_target,
                    None => continue;
                };
```

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+                    // compare the hashes of the target from the repo and the
+                    // target we just created. If they are the same, this must
+                    // be the same file, symlink it.
+                    if target_from_path.hashes.sha256 == repo_target.hashes.sha256 {
```
Similar to above suggestions, remove some nesting:
```suggestion
                    ensure!(target_from_path.hashes.sha256 == repo_target.hashes.sha256,
                        error::HashMismatch { bla });
```

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+                        symlink(entry.path(), &dest).context(error::LinkCreate { path: &dest })?;
+                    } else {
+                        return error::HashMismatch {
+                            context: "target".to_string(),
```
```suggestion
                            context: "target",
```
Snafu should do this, and only when it needs to (though this is currently in an error branch so it'd only happen on error... still, no need to do it manually)

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+                    if target_from_path.hashes.sha256 == repo_target.hashes.sha256 {
+                        let dest = if self.root.signed.signed.consistent_snapshot {
+                            abs_outdir.join(format!(
+                                "{}.{}",
```
Similar to the json file naming, we're now doing this name generation in 4 places, should we standardize?  (Repository::target_digest_and_filename, tuftool::metadata::write_metadata, tuftool::create::process_target, and now this)

zmrow replied:
https://github.com/awslabs/tough/issues/126

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+/// `RepositoryEditor` contains the various bits of data needed to construct
+/// or edit a TUF repository.
+///
+/// A new repository may be created using the `new()` method.
```
```suggestion
/// A new repository may be started using the `new()` method.
```

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+/// Targets, versions, and expirations may be added to their respective roles
+/// via the provided "setter" methods. The final step in the process is the
+/// `sign()` method, which takes a given set of signing keys, builds each of
+/// the roles using the data provided, and signs the roles.
```
```suggestion
/// the roles using the data provided, and signs the roles.
/// This results in a `SignedRepository` which can be used to write the repo to disk.
```

tjkirch commented on tough/src/error.rs:
```diff
@@ -176,12 +230,42 @@ pub enum Error {
+    SigningKeysNotFound { role: String },
+
+    #[snafu(display(
+        "Tried to load repository with spec version '{}', version '{}' is supported",
```
Not necessarily a repository; we also use this error when we try to set targets, which you can do in a fresh editor.

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+        Ok(editor)
+    }
+
+    /// Add an existing `Targets` struct to the repository. Any existing
+    /// targets are preserved.
```
Not sure I'm understanding this - existing targets aren't preserved, we're fully replacing `self.existing_targets`.  Should this be called something like `set_targets` or `replace_targets`?  Or were you intending to merge the given targets with `existing_targets`?

tjkirch replied:
Another option is having `add_targets` and `clear_targets` methods, so a user can work with `Targets` objects if desired (merging, replacing, etc.), and changing the path-based methods below to `add_target(s)_path` if they want to work with paths..?  One use case we've discussed is merging repos, so this distinction seems useful.

zmrow replied:
Good point. I'll clarify the comment as well.

One method I completely missed is a `remove_target` method as well.

zmrow replied:
Oh, wasn't thinking when I answered initially. When a user passes a `Targets` struct to this method, we preserve any `Target` objects in `self.existing_targets`.  When a user adds targets using the `add_target(s)()` methods, they are saved to `self.new_targets`. We combine the two maps when creating the final repository.

I agree that a `clear_targets()` method would be a good idea.

I'm wondering how we best support a `remove_target()` method. The user would need to supply a target filename, since we store this information as `HashMap<filename, Target_object>`

tjkirch replied:
Could wait for a use case / request, but I'd assume they probably have a filename/path, so it could build a Target based on that and remove it from existing/new lists...  doesn't seem urgent if we have clear.

zmrow replied:
Agreed. I'll implement `clear_targets()` for now.

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+        Ok(self)
+    }
+
+    /// Add an existing `Snapshot` to the repository. Only the `_extra` data
+    /// is preserved
```
If the intent is just to allow setting `snapshot_extra`, and we always construct `snapshot`, should we call this `snapshot_extra`?

(Same with `fn timestamp`)

zmrow replied:
I see your point. The `*_extra` member of the role structs only exist to capture extra arguments found during deserialization and not something I'd think a user would probably interact with in most cases.

I really see these `snapshot` and `timestamp` setters as convenience methods so a user can just pass the entire object. They keep a consistent API, allow us to check the spec version, and give us the opportunity to do more in the future if the TUF spec changes.

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+
+        // Hold on to the existing targets
+        self.existing_targets = Some(targets.targets);
+        self.targets_extra = Some(targets._extra);
```
Do we need methods to set *_extra fields?  Currently these methods that fully replace exiting targets/bla are the only way to set it.

zmrow replied:
I'm not opposed to it, but I'm wondering how much of a use case it will really have.

tjkirch replied:
I agree, I'd rather leave it out.  I hope no one would rely on those extra fields just because they can set them when deserializing.  (I hope we can eliminate them at some point...)

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+        })?;
+        let _extra = self.snapshot_extra.clone().unwrap_or_else(HashMap::new);
+
+        let targets_meta = Self::snapshot_meta(signed_targets);
```
Sorry if this is a dumb question... why do we use `snapshot_meta` to make a variable called `targets_meta` associated with `targets.json`, and below we use `timestamp_meta` to make a variable called `snapshot_meta` associated with `snapshot.json`?  Naively they seem to be used backwards... :)

I think we could use a little more documentation on/in these build methods.  It's complicated stuff.

zmrow replied:
Not a dumb question at all. I think the confusion comes from the variable name (suggestions welcome). Snapshot contains the filename and hashes of the signed `Targets` and `Root` data. Further details below.

The `Snapshot` struct has a member `meta`. `Snapshot.meta` contains a `HashMap<String, SnapshotMeta>`.  This `Hashmap` is where the data about the `Targets/Root` is stored. In practice it looks something like:

```
{
   ...,
   "meta":{
      "root.json":{
         "length":1,
         "hashes":{
            "sha256":"1234"
         },
         "version":1
      },
      "targets.json":{
         "length":2,
         "hashes":{
            "sha256":"5678"
         },
         "version":1588449897
      }
   }
}
```

tjkirch replied:
Maybe a comment above this like `// Snapshot stores metadata about targets and root.`  And below in the other function, `// Targets stores metadata about snapshot.` or something..?

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+        }
+    }
+
+    // Signs a given role. This function finds the appropriate keys to sign
+    // given role, signs the role, and builds a `SignedRole<T>`.
```
Can we use docstrings `///` instead of comments?

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+
+        let targets = editor.existing_targets.as_ref().unwrap();
+
+        assert!(editor.targets_version.is_none());
```
I'm not sure of the end goal of this test function, other than the unwraps not failing - these `is_none` checks don't seem to be helping us much since that'd be the default :)  I guess it checks that file1/file2 are in targets?  Is that enough to have confidence?

webern replied:
> I'm not sure of the end goal of this test function, other than the unwraps not failing - these `is_none` checks don't seem to be helping us much since that'd be the default :) I guess it checks that file1/file2 are in targets? Is that enough to have confidence?

The issue might be that we are using the simplistic 'reference implementation' tuf repo. We should consider adding a more complex tuf repo to the test data. One that has consistent snapshots and multiple keys, for example. In other words, the assertions are as boring as the tuf repo is.

Further: We should develop the test utils to build repos on-the-fly during testing to start doing more varied testing with edge cases and such.

zmrow replied:
I guess I saw this as more of a "unit" test for the methods that add data from existing roles. I just want to make sure that we only keep the data we want to.

tjkirch reviewed (commented):
I had a hard time following the build logic, so I think that could use some documentation, but the code structure seems clear overall.

webern commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,660 @@
+
+impl SignedRepository {
+    /// Writes the metadata to the given directory. If consistent snapshots
+    /// are used, the appropriate files are prefixed with their `sha256`.
```
slightly confusing. in the snapshot.json file these sha256 prefixes should appear, but the metadata files get a version prefix.

zmrow replied:
Oops - you're right. Will fix.

webern commented on tough/src/schema/mod.rs:
```diff
@@ -267,12 +301,28 @@ impl Target {
 }
 
+impl Targets {
+    pub fn new(spec_version: String, version: NonZeroU64, expires: DateTime<Utc>) -> Self {
```
These `new` functions probably address @iliana's concern in #20

zmrow replied:
I don't currently implement `new()` for all of these types unfortunately. :(

webern reviewed (approved):
🍿

zmrow commented:
Addressed almost all of @tjk's and @webern's comments.

Open items:

- [ ] Refactor a bit and split out the signed* types to better emphasize the main method of editor (`sign()`)

- [ ] Further editing of targets:
  * `clear_targets()`: Clear both targets HashMaps.
  * ~~What is the best interface for "removing" a single target from the list? By path? By filename? (preferred)~~ (Skipping for now)

- [ ] Better integration test. Create/edit/write/read. (Currently working on this one)

- [x] Unifying the file naming logic - outside the scope of this PR. Will create an issue and link here. --> https://github.com/awslabs/tough/issues/126

tjkirch commented:
> What is the behavior of `clear_targets()`? Will it clear the targets added from the Targets object? or targets added by path? Differentiating sounds weird and has a crappy interface.

I assumed it would set new and existing to `Some(HashMap::new())`.  Really clear all targets.  No?

> What is the best interface for "removing" a single target from the list? By path? By filename? (preferred)

I think we could/should skip this for now; we don't have a use case or a clear choice.

> Open items:

I'm also hoping we can split up editor.rs a bit, as we discussed offline - the signed* types seem like an easy split, and we could emphasize the "main" method of the editor, `sign`, more easily then.

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,744 @@
+        std::fs::create_dir_all(outdir).context(error::DirCreate { path: outdir })?;
+        // Get the absolute path of the indir and outdir
+        let abs_indir = std::fs::canonicalize(&indir).context(error::AbsolutePath {
+            path: indir.as_ref().to_owned(),
```
```suggestion
            path: indir.as_ref(),
```
Shouldn't need that, same as outdir, right?

zmrow replied:
Good catch - I missed that one.

tjkirch commented on tough/src/editor.rs:
```diff
@@ -0,0 +1,744 @@
+
+            // If the entry is a file...
+            if !entry.file_type().is_file() {
+                continue;
+            } else {
```
```suggestion
                continue;
            }
```
You don't need all the code below to be indented for this.  The `continue` will move to the next iteration and skip it all.

zmrow replied:
😳 

Sorry about that... This is a silly mistake.

zmrow commented:
You're right about the split, I was gonna introduce that later. :) I edited my comment to reflect the work.

zmrow commented:
This push addresses the remaining open items except for the refactor into separate files. I'll do that in another push to make it easier to review.

- [ ] Refactor a bit and split out the signed* types to better emphasize the main method of editor (`sign()`)

- [x] Further editing of targets:
  * `clear_targets()`: Clear both targets HashMaps.
  * ~~What is the best interface for "removing" a single target from the list? By path? By filename? (preferred)~~ (Skipping for now)

- [x] Better integration test. Create/edit/write/read.

- [x] Unifying the file naming logic - outside the scope of this PR. Will create an issue and link here. --> https://github.com/awslabs/tough/issues/126

zmrow commented:
Appeased clippy... angry little fellow he is...  :)

tjkirch reviewed (approved):
LGTM - hoping to get the file split discussed above so that the editor code stands out clearly, but I'm sure that'll work fine.

zmrow commented:
This push is a refactor into separate modules as per @tjkirch 's request.  It's much nicer now. The only change to the code is that I moved the `RepositoryEditor::sign_role()` method to `SignedRole::new()` which feels a bit more ergonomic. 

To wrap close the loop:
- [x] Refactor a bit and split out the signed* types to better emphasize the main method of editor (`sign()`)

- [x] Further editing of targets:
  * `clear_targets()`: Clear both targets HashMaps.
  * ~~What is the best interface for "removing" a single target from the list? By path? By filename? (preferred)~~ (Skipping for now)

- [x] Better integration test. Create/edit/write/read.

- [x] Unifying the file naming logic - outside the scope of this PR. Will create an issue and link here. --> https://github.com/awslabs/tough/issues/126

webern reviewed (approved):
🎪

zmrow commented:
No-op push to kick the checks again
